doc/supervisor-trees.md

```diff
@@ -0,0 +1,58 @@
+# Supervisor trees in GD2
+
+GD2 uses supervisor trees to manage its internal services, which provides a clean and consistent pattern to implement these services.
+
+GD2 uses the [`suture`][1] package to implement its supervisor tree, which was inspired by the Erlang/Elixir supervisor trees. More information on supervisor trees and the `suture` package can be found in [this blog post][2].
+
+## Quick introduction
+
+Supervisor trees involve 2 types of objects, supervisors and services.
+
+### Services
+
+Services are objects that provide or perform actual services. Services are managed by supervisors, and can be added to only one supervisor at a time. Services are started and managed by supervisors.
+
+### Supervisors
+
+Supervisors are a specialized services, which manage other services. Supervisors manage the lifecyle of the services that have been added to them. They take care of starting, restarting, stopping and keeping the services alive.
+
+Supervisors themselves can be added as services to other supervisors. This way supervisor trees can be built to manage lots of different services in the application.
+
+Only the top-most (root) supervisor in a supervisor tree is started manually. All other supervisors will be started by their parent supervisors.
+
+
+## The GD2 supervisor tree
+
+Currently, supervisor trees are to manage the network services provides by GD2 (rest, grpc, sunrpc). In the future, services like etcd, plugins etc. can also be managed using supervisor trees.
+
+The strucutre of the GD2 supervisor tree currently is as shown below.
+
+```
+
+  gd2-main
+  |
+  +-->gd2-servers
+      |
+      +-->peerrpc
+      |
+      +-->gd2-muxserver
+          |
+          |-->muxlistener
+          |
+          |-->rest
+          |
+          +-->sunrpc
+
```

Comment:
Please add a note on what order these are stopped and restarted. To-down/bottom-up ?